### PR TITLE
[#57] Backend as a user I can login with an Oauth provider

### PR DESCRIPTION
**Pull request by phisith**
https://github.com/nimblehq/js-framework-benchmark/issues/57

## What happened 👀

- Setup authentication (login using Google), 
- Save user information into database when login
- Add login url and protected route

## Insight 📝

To be able to setup the authentication I'm using the [remix-auth](https://www.npmjs.com/package/remix-auth) with [remix-auth-socials](https://www.npmjs.com/package/remix-auth-socials) (A collection of Remix Auth strategies for Oauth2 Social logins.).

In this setup I'm not able to use the `Passport.js`, due to Remix doesn't have the ability to control the server code like Nextjs. 

Reference:
Screenshot 2566-05-23 at 17 58 36

### Login Flow
https://github.com/nimblehq/js-framework-benchmark/assets/59072804/4f705730-4e75-462b-9168-73e26f2f49cf

### session is store in cookies
Screenshot 2566-05-24 at 09 22 03

### Protected route (must login before access the route)
Screenshot 2566-05-23 at 16 04 54

### CODE REVIEW

**Review by tednguyendev — APPROVED** [review]

**Review by olivierobert — CHANGES_REQUESTED** [review]

File: remix/app/services/session.server.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { createCookieSessionStorage } from '@remix-run/node';
+
+export const sessionStorage = createCookieSessionStorage({
```
[the remark this answers is not in the document]
Comment (phisith) [reply remix/app/services/session.server.ts:3]: Oh, I just forgot to remove it when I refactor the session.server 🙊.

Now removed.

File: remix/app/routes/_index.tsx [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
-export const meta: V2_MetaFunction = () => {
-  return [{ title: "Remix App Benchmark ⚔" }];
-};
+import { Form } from '@remix-run/react';
+import { SocialsProvider } from 'remix-auth-socials';
 
 export default function Index() {
   return (
-    <div style={{ fontFamily: "system-ui, sans-serif", lineHeight: "1.4" }}>
+    <div style={{ fontFamily: 'system-ui, sans-serif', lineHeight: '1.4' }}>
```
[the remark this answers is not in the document]
Comment (phisith) [reply remix/app/routes/_index.tsx]: Well noted, Actually this styling is come with the default setup, so next time I will pay attention to it 👍🏼.

Removed.

File: remix/app/repositories/user.server.test.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import type { User } from '@prisma/client';
+
+import UserRepository from '../repositories/user.server';
+import { userFactory } from '../tests/factories/user.factory';
+import { prismaMock } from '../tests/setup-test-env';
+
+describe('User Repository', () => {
+  describe('updateOrCreate', () => {
+    describe('given there is a user information', () => {
```
[the remark this answers is not in the document]
Comment (phisith) [reply remix/app/repositories/user.server.test.ts:9]: Fixed.

File: remix/app/services/auth.server.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import UserRepository from '../repositories/user.server';
+import type { UserProfile } from '../types';
+
+interface Profile {
+  profile: { _json: UserProfile };
+}
+export async function handleSocialAuthCallBack({ profile }: Profile) {
+  const userProfile: UserProfile = profile?._json;
+
+  if (userProfile) {
```
[the remark this answers is not in the document]
Comment (phisith) [reply remix/app/services/auth.server.ts:10]: For this one and the [question](https://github.com/nimblehq/js-framework-benchmark/pull/90#discussion_r1208982482) above is I have to return the profile for `remix-auth` as it require for the callback.

Short answer is: Yes, it handle by `remix-auth`.

File: remix/app/services/auth.server.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import UserRepository from '../repositories/user.server';
+import type { UserProfile } from '../types';
+
+interface Profile {
+  profile: { _json: UserProfile };
+}
```
[the remark this answers is not in the document]
Comment (phisith) [reply remix/app/services/auth.server.ts:8]: Added.

File: remix/app/routes/auth.google.callback.tsx [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import type { LoaderArgs } from '@remix-run/node';
+import { SocialsProvider } from 'remix-auth-socials';
+
+import { authenticator } from '../config/auth.server';
+
+export const loader = ({ request }: LoaderArgs) => {
+  return authenticator.authenticate(SocialsProvider.GOOGLE, request, {
+    successRedirect: '/newsletters',
+    failureRedirect: '/',
```
[the remark this answers is not in the document]
Comment (phisith) [reply remix/app/routes/auth.google.callback.tsx:9]: For this I actually done it same as you prefer in PR[#58], should I change here too or wait that PR merge ?

File: remix/app/routes/newsletters.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import type { User } from '@prisma/client';
+import type { LoaderArgs } from '@remix-run/node';
+import { Outlet, useLoaderData } from '@remix-run/react';
+
+import { authenticator } from '../config/auth.server';
+import UserRepository from '../repositories/user.server';
+import type { UserProfile } from '../types';
+
+export const loader = async ({ request }: LoaderArgs) => {
+  const profile = (await authenticator.isAuthenticated(request)) as {
```
[the remark this answers is not in the document]
Comment (phisith) [reply remix/app/routes/newsletters.tsx:10]: In this point I also have done it in PR [#58], But I will change here too.

File: remix/app/routes/auth.google.callback.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (olivierobert) [reply remix/app/routes/auth.google.callback.tsx:9]: This can wait then the next PR then :-)

**Review by olivierobert — APPROVED** [review]

**Pull request merged**